aten/src/ATen/core/LegacyTypeDispatch.h

```diff
@@ -50,7 +50,34 @@ CAFFE2_API LegacyTypeDispatch& globalLegacyTypeDispatch();
 // A RAII, thread local (!) guard that will disable dispatch to variable
 // handler.
 //
-// See NOTE [ Treating Variables as non-Variables in type dispatch ] for details.
+// NOTE [ Treating Variables as non-Variables in type dispatch ]
+//
+// What exactly does AutoNonVariableType do?  The short answer is, it causes
+// dispatches on ATen functions to go to the non-variable implementation,
+// bypassing autograd handling (and also profiling and tracing).
+//
+// To understand why this guard exists, it's helpful to understand the history
+// behind how Variable was implemented.  Previously, Variables were implemented
+// as a wrapper on Tensors; so the act of processing a Variable involved
+// unwrapping the underlying Tensor, and then calling the underlying base
+// operation on /that/ operation
+//
+// However, after the Variable/Tensor merge, there is no concept of unwrapping
+// a tensor anymore.  If you just call the operation on the same variable
+// again inside your VariableType handler, you'll dispatch back to
+// VariableType, which is not what we want.
+//
+// The solution to the above problem is to add `at::NonVariableTypeMode`, which
+// when enabled will cause `legacyTensorType()` and `getType()` to always return
+// non-Variable type, even if the tensor being called on is a variable.
+//
+// TODO: Since `torch::NoGradGuard` serves the same purpose in libtorch, we should
```

Comment:
I'd say "serves _almost_ the same purpose" or something -- this is something we got wrong before and the statement as it is now is definitely misleading.